File: server/src/services/GuestService.ts
import { Guest } from "../models/Guest";
import Logger from "../util/logs/logger";
import { Readable } from 'stream';
import csvParser from 'csv-parser';
import ExcelJS from 'exceljs';
import { inject, injectable } from "inversify";
import { SubEventDao } from "../dao/SubEventDao";
import { TYPES } from "../configs/types";
import { SubEvent } from "../models/SubEvent";
import { GuestDao } from "../dao/GuestDao";
import { UploadValidation } from "../models/UploadValidation";
import { DUPLICATE_GUEST, EMAIL_BAD_FORMAT, EMAIL_MISSING, EMAIL_TOO_LONG, MISSING_EVENT, NAME_MISSING, NAME_TOO_LONG, PARTY_NUMBER_MISSING, PHONE_TOO_LONG } from "../util/upload/ValidationErrors";
import validator from 'validator';
import { TransportationType } from "../models/TransportationType";
import { Transportation } from "../models/Transportation";
import { Accommodation } from '../models/Accommodation';
import { Drinks } from "../models/Drinks";
import { Parser } from 'json2csv';
import { CustomError } from "../util/error/CustomError";
import { UploadGuestLists } from "../models/UploadGuestLists";
import { AccommodationDao } from "../dao/AccommodationDao";

@injectable()
export class GuestService {

    constructor(
        @inject(TYPES.SubEventDao) private subEventDao: SubEventDao,
        @inject(TYPES.GuestDao) private guestDao: GuestDao,
        @inject(TYPES.AccommodationDao) private accommodationDao: AccommodationDao
    ) { }

    async parseCSV(buffer: Buffer, eventId: string): Promise<UploadGuestLists> {
        const createGuests: Guest[] = [];
        const updateGuests: Guest[] = [];
        const stream = Readable.from(buffer.toString());

        const subEvents: SubEvent[] = await this.subEventDao.getAllSubEvents(eventId);
        const accommodations: Accommodation[] = await this.accommodationDao.getAllAccommodations(eventId);

        return new Promise((resolve, reject) => {
            stream.pipe(csvParser())
                .on('data', (data) => {
                    if (data['Id']) {
                        const guest = this.updateGuestFromData(data, accommodations);
                        updateGuests.push(guest);
                    } else {
                        const guest = this.createGuestFromData(data, eventId, subEvents, accommodations);
                        createGuests.push(guest);
                    }
                })
                .on('end', () => {
                    resolve({ createGuests, updateGuests } as UploadGuestLists);
                })
                .on('error', (error) => {
                    reject(error);
                });
        });
    }

    async parseExcel(buffer: Buffer, eventId: string): Promise<UploadGuestLists> {
        const createGuests: Guest[] = [];
        const updateGuests: Guest[] = [];
        const workbook = new ExcelJS.Workbook();
        await workbook.xlsx.load(buffer);
        const worksheet = workbook.worksheets[0];

        const headers = worksheet.getRow(1).values as string[];

        const subEvents: SubEvent[] = await this.subEventDao.getAllSubEvents(eventId);
        const accommodations: Accommodation[] = await this.accommodationDao.getAllAccommodations(eventId);

        worksheet.eachRow({ includeEmpty: false }, (row, rowNumber) => {
            if (rowNumber === 1) return;

            const data: { [key: string]: any } = {};
            const rowValues = row.values as (string | number | undefined)[];

            headers.forEach((header, index) => {
                const value = rowValues[index];
                data[header] = value;
            });

            if (data['Id']) {
                const guest = this.updateGuestFromData(data, accommodations);
                updateGuests.push(guest);
            } else {
                const guest = this.createGuestFromData(data, eventId, subEvents, accommodations);
                createGuests.push(guest);
            }

        });

        return { createGuests, updateGuests } as UploadGuestLists;
    }

    private createGuestFromData(data: any, eventId: string, subEvents: SubEvent[], accommodations: Accommodation[]): Guest {
        const guest: Partial<Guest> = {};

        guest.id = '';
        guest.eventId = eventId;
        guest.name = (data['Name'] || '').trim();
        guest.email = (data['Email'] || '').trim();
        guest.phone = this.cleanPhoneField(data['Phone']);
        guest.serialNumber = (data['Serial Number'] || '').trim();
        guest.groupNumber = Number(data['Party Number'] || -1);
        guest.dietaryRestrictions = (data['Dietary Restrictions'] || '').trim();

        // SubEvents
        guest.subEvents = this.parseGuestEvents(data, subEvents);

        // Arrival Transportation
        guest.arrival = this.parseTransportation(data, 'Arrival Type', true);

        // Departure Transportation
        guest.departure = this.parseTransportation(data, 'Departure Type', false);

        // Accommodation
        guest.accommodation = this.parseAccommodation(data, accommodations);

        return guest as Guest;
    }

    private updateGuestFromData(data: any, accommodations: Accommodation[]): Guest {
        const guest: Partial<Guest> = {};

        guest.id = data['Id'].trim();
        guest.eventId = data['Event Id'];
        guest.name = (data['Name'] || '').trim();
        guest.email = (data['Email'] || '').trim();
        guest.serialNumber = (data['Serial Number'] || '').trim();
        guest.groupNumber = Number(data['Party Number'] || -1);
        guest.dietaryRestrictions = (data['Dietary Restrictions'] || '').trim();

        // Arrival Transportation
        guest.arrival = this.parseTransportation(data, 'Arrival Type', true);

        // Departure Transportation
        guest.departure = this.parseTransportation(data, 'Departure Type', false);

        // Accommodation
        guest.accommodation = this.parseAccommodation(data, accommodations);

        return guest as Guest;
    }

    private parseGuestEvents(data: any, subEvents: SubEvent[]): SubEvent[] {
        const events: SubEvent[] = [];

        for (let weddingEvent of subEvents) {
            const invitedStatus = data[weddingEvent.name];
            if (invitedStatus && ['yes', 'done', 'y', 'true'].includes(String(invitedStatus).trim().toLowerCase())) {
                events.push(weddingEvent);
            }
        }

        return events;
    }

    private parseTransportation(data: any, typeField: string, isArrival: boolean): Transportation | undefined {
        const transportationTypeValue = data[typeField];
        if (!transportationTypeValue) return undefined;

        const prefix = isArrival ? "Arr" : "Dep";

        let transportation = {
            isArrival: isArrival,
            type: transportationTypeValue,
            number: String(data[`${prefix}.Num`]).trim()
        } as Transportation;

        transportation.time = this.combineDateTime(data[`${prefix}.Date`], data[`${prefix}.Time`]);

        return transportation;
    }

    private combineDateTime(dateStr: string, timeStr: string): string {
        const date = new Date(`${dateStr} ${timeStr}`);

        const yyyy = date.getFullYear();
        const mm = String(date.getMonth() + 1).padStart(2, '0');
        const dd = String(date.getDate()).padStart(2, '0');
        const hh = String(date.getHours()).padStart(2, '0');
        const min = String(date.getMinutes()).padStart(2, '0');
        const ss = String(date.getSeconds()).padStart(2, '0');

        return `${yyyy}-${mm}-${dd}T${hh}:${min}:${ss}`;

    }

    private cleanPhoneField(phone: string): string {
        if (!phone) {
            return '';
        }
        let cleaned = phone.trim().replace(/[^\d+]/g, '');
        if (cleaned.indexOf('+') > 0) {
            cleaned = cleaned.replace(/\+/g, '');
        }
        return cleaned;
    }

    private parseAccommodation(data: any, accommodations: Accommodation[]): Accommodation {

        const accommodationName = data['Accommodation'];
        const roomNumber = data['Room Number'];

        const assignedAccommodation = accommodations.find(acc => acc.name === accommodationName);

        let accommodation = {} as Accommodation;

        if (assignedAccommodation) {
            accommodation = assignedAccommodation;
            accommodation.roomNumber = roomNumber;
        }

        return accommodation;
    }


    async validateGuests(eventId: string, guestLists: UploadGuestLists): Promise<UploadValidation> {
        const currentGuestNames = (await this.guestDao.getAllGuests(eventId)).map(guest => guest.name.toLowerCase());

        const uploadIssues: Map<string, string> = new Map<string, string>();
        const validatedCreateGuests: Guest[] = [];
        const validatedUpdateGuests: Guest[] = [];

        guestLists.createGuests.forEach((guest, index) => {
            const validationError = this.validateGuest(guest, currentGuestNames);
            if (validationError) {
                uploadIssues.set(guest.name || `Unknown ${index}`, validationError);
            } else {
                validatedCreateGuests.push(guest);
            }
        });

        guestLists.updateGuests.forEach((guest, index) => {
            const validationError = this.validateGuest(guest, currentGuestNames, true);
            if (validationError) {
                uploadIssues.set(guest.name || `Unknown ${index}`, validationError);
            } else {
                validatedUpdateGuests.push(guest);
            }
        });

        const uploadGuestLists = new UploadGuestLists(validatedCreateGuests, validatedUpdateGuests);

        return new UploadValidation(uploadIssues, uploadGuestLists);
    }

    private validateGuest(guest: Guest, currentGuestNames: string[], isUpdate = false): string | null {
        if (validator.isEmpty(guest.name)) {
            return NAME_MISSING;
        }

        if (guest.groupNumber === -1) {
            return PARTY_NUMBER_MISSING;
        }

        if (!isUpdate && currentGuestNames.includes(guest.name.toLowerCase())) {
            return DUPLICATE_GUEST;
        }

        if (guest.name.length > 100) {
            return NAME_TOO_LONG;
        }

        if (guest.email && guest.email.length > 50) {
            return EMAIL_TOO_LONG;
        }

        if (guest.phone && guest.phone.length > 50) {
            return PHONE_TOO_LONG;
        }

        return null;
    }

    private csvFields = [
        { label: 'Id', value: 'id' },
        { label: 'Wedding Id', value: 'eventId' },
        { label: 'Name', value: 'name' },
        { label: 'Party Number', value: 'groupNumber' },
        { label: 'Serial Number', value: 'serialNumber' },
        { label: 'SubEvents', value: 'subevents' },
        { label: 'Email', value: 'email' },
        { label: 'Accommodation', value: 'accommodation.name' },
        { label: 'Room Number', value: 'accommodation.roomNumber' },
        { label: 'Dietary Restrictions', value: 'dietaryRestrictions' },
        { label: 'Type of Drink', value: 'drinks.preferences' },
        { label: 'Arrival Type', value: 'arrival.type' },
        { label: 'Arr.Date', value: 'arrival.date' },
        { label: 'Arr.Time', value: 'arrival.time' },
        { label: 'Arr.Num', value: 'arrival.number' },
        { label: 'Departure Type', value: 'departure.type' },
        { label: 'Dep.Date', value: 'departure.date' },
        { label: 'Dep.Time', value: 'departure.time' },
        { label: 'Dep.Num', value: 'departure.number' },
    ];

    generateCSVContent(guests: Guest[]): string {
        try {
            const processedGuests = guests.map(guest => {
                const processedGuest = { ...guest } as any;

                // Format Arrival
                const guestArrival = guest.arrival as any;
                if (guestArrival) {
                    const [arrDate, arrTime] = guestArrival.time ? guestArrival.time.split('T') : ['', ''];
                    processedGuest.arrival = {
                        ...guestArrival,
                        date: arrDate ? this.formatDateString(arrDate) : '',
                        time: arrTime || '',
                    };
                }

                // Format Departure
                const guestDeparture = guest.departure as any;
                if (guestDeparture) {
                    const [depDate, depTime] = guestDeparture.time ? guestDeparture.time.split('T') : ['', ''];
                    processedGuest.departure = {
                        ...guestDeparture,
                        date: depDate ? this.formatDateString(depDate) : '',
                        time: depTime || '',
                    };
                }

                if (guest.drinks && Array.isArray(guest.drinks.preferences)) {
                    processedGuest.drinks = {
                        ...guest.drinks,
                        preferences: guest.drinks.preferences.join(', '),
                    };
                }

                if (guest.subEvents && Array.isArray(guest.subEvents)) {
                    processedGuest.subEvents = guest.subEvents.map(subEvent => subEvent.name).join(', ');
                }
                return processedGuest;
            });

            const opts = {
                fields: this.csvFields,
                header: true,
                quote: '"'
            };

            const parser = new Parser(opts);
            const csv = parser.parse(processedGuests);

            return csv;
        } catch (error: any) {
            throw new CustomError(error.message, 500);
        }
    }

    private formatDateString(dateStr: string): string {
        const [year, month, day] = dateStr.split('-');
        return `${month}/${day}/${year}`;
    }


}

File: server-go/internal/guests/guest_service.go
package guests

import (
	"encoding/csv"
	"fmt"
	"io"
	"mime/multipart"
	"regexp"
	"strconv"
	"strings"
	"time"
	"weddingpass/server/internal/accommodation"
	"weddingpass/server/internal/common"
	"weddingpass/server/internal/drinks"
	"weddingpass/server/internal/logger"
	"weddingpass/server/internal/subevents"
	"weddingpass/server/internal/transportation"

	"github.com/xuri/excelize/v2"
)

type GuestService interface {
	common.IBaseService[*GuestDTO]
	FetchPartyMembers(params map[string]string) ([]*GuestDTO, *common.CustomError)
	GetGuestByName(params map[string]string) (*GuestDTO, *common.CustomError)
	GetGuestsByPhone(params map[string]string) ([]*GuestDTO, *common.CustomError)
	GetGuestsByEmail(params map[string]string) ([]*GuestDTO, *common.CustomError)
	GetGuestBySerialNumber(params map[string]string) (*GuestDTO, *common.CustomError)
	GetGuestsForSubEvent(params map[string]string) ([]*GuestDTO, *common.CustomError)
	GetGuestsByHotel(params map[string]string) ([]*GuestDTO, *common.CustomError)
	ValidateGuests(params map[string]string, guestLists *UploadGuestListsDTO) (*UploadValidationDTO, *common.CustomError)
	UploadGuests(params map[string]string, file multipart.File, ext string) (*UploadGuestListsDTO, *common.CustomError)
	GenerateCSVContent(params map[string]string) (string, *common.CustomError)
}

type guestService struct {
	common.BaseService[*Guest, *GuestDTO]
	GuestRepository      GuestRepository
	SubEventService      subevents.SubEventService
	AccommodationService accommodation.AccommodationService
	Converter            *GuestConverter
}

func NewGuestService(repository GuestRepository,
	transportationConverter *transportation.TransportationConverter,
	subEventConverter *subevents.SubEventConverter,
	subEventService subevents.SubEventService,
	drinksConverter *drinks.DrinksConverter,
	accommodationConverter *accommodation.AccommodationConverter,
	accommodationService accommodation.AccommodationService) GuestService {
	converter := NewGuestConverter(transportationConverter, subEventConverter, subEventService, drinksConverter, accommodationConverter)
	base := common.BaseService[*Guest, *GuestDTO]{
		Repository: repository,
		ConvertToDAO: func(guest *GuestDTO) (*Guest, *common.CustomError) {
			return converter.ConvertGuestToDAO(guest)
		},
		ConvertToDTO: func(rawGuest *Guest) (*GuestDTO, *common.CustomError) {
			return converter.ConvertGuestToDTO(rawGuest)
		},
	}
	return &guestService{
		BaseService:          base,
		GuestRepository:      repository,
		SubEventService:      subEventService,
		AccommodationService: accommodationService,
		Converter:            converter,
	}
}

func (s *guestService) FetchPartyMembers(params map[string]string) ([]*GuestDTO, *common.CustomError) {
	rawGuests, err := s.GuestRepository.FetchPartyMembers(params)
	if err != nil {
		return nil, err
	}
	guestDtos, err := s.ConvertAllToDTO(rawGuests)
	if err != nil {
		return nil, err
	}
	return guestDtos, nil
}

func (s *guestService) GetGuestByName(params map[string]string) (*GuestDTO, *common.CustomError) {
	rawGuest, err := s.GuestRepository.GetGuestByName(params)
	if err != nil {
		return nil, err
	}
	guestDto, err := s.ConvertToDTO(rawGuest)
	if err != nil {
		return nil, err
	}
	return guestDto, nil
}

func (s *guestService) GetGuestsByPhone(params map[string]string) ([]*GuestDTO, *common.CustomError) {
	rawGuests, err := s.GuestRepository.GetGuestsByPhone(params)
	if err != nil {
		return nil, err
	}
	guestDtos, err := s.ConvertAllToDTO(rawGuests)
	if err != nil {
		return nil, err
	}
	return guestDtos, nil

}

func (s *guestService) GetGuestsByEmail(params map[string]string) ([]*GuestDTO, *common.CustomError) {
	rawGuests, err := s.GuestRepository.GetGuestsByEmail(params)
	if err != nil {
		return nil, err
	}
	guestDtos, err := s.ConvertAllToDTO(rawGuests)
	if err != nil {
		return nil, err
	}
	return guestDtos, nil
}

func (s *guestService) GetGuestBySerialNumber(params map[string]string) (*GuestDTO, *common.CustomError) {
	rawGuest, err := s.GuestRepository.GetGuestBySerialNumber(params)
	if err != nil {
		return nil, err
	}
	guestDto, err := s.ConvertToDTO(rawGuest)
	if err != nil {
		return nil, err
	}
	return guestDto, nil
}

func (s *guestService) GetGuestsForSubEvent(params map[string]string) ([]*GuestDTO, *common.CustomError) {
	rawGuests, err := s.GuestRepository.GetGuestsForSubEvent(params)
	if err != nil {
		return nil, err
	}
	guestDtos, err := s.ConvertAllToDTO(rawGuests)
	if err != nil {
		return nil, err
	}
	return guestDtos, nil
}

func (s *guestService) GetGuestsByHotel(params map[string]string) ([]*GuestDTO, *common.CustomError) {
	rawGuests, err := s.GuestRepository.GetGuestsByHotel(params)
	if err != nil {
		return nil, err
	}
	guestDtos, err := s.ConvertAllToDTO(rawGuests)
	if err != nil {
		return nil, err
	}
	return guestDtos, nil
}

func (s *guestService) ValidateGuests(params map[string]string, guestLists *UploadGuestListsDTO) (*UploadValidationDTO, *common.CustomError) {
	existingGuests, err := s.GetAll(params)
	if err != nil {
		return nil, err
	}
	var existingGuestNames []string
	for _, existingGuest := range existingGuests {
		existingGuestNames = append(existingGuestNames, strings.ToLower(existingGuest.Name))
	}
	var uploadIssues = make(map[string]string)
	var validatedCreateGuests, validatedUpdateGuests []GuestDTO

	for index, createGuest := range guestLists.CreateGuests {
		validationError := validateGuest(createGuest, existingGuestNames, false)
		if validationError != "" {
			name := createGuest.Name
			if name == "" {
				name = fmt.Sprintf("Unknown %d", index)
			}
			uploadIssues[name] = validationError
		} else {
			validatedCreateGuests = append(validatedCreateGuests, createGuest)
		}
	}

	for index, updatedGuest := range guestLists.UpdateGuests {
		validationError := validateGuest(updatedGuest, existingGuestNames, true)
		if validationError != "" {
			name := updatedGuest.Name
			if name == "" {
				name = fmt.Sprintf("Unknown %d", index)
			}
			uploadIssues[name] = validationError
		} else {
			validatedUpdateGuests = append(validatedUpdateGuests, updatedGuest)
		}
	}

	uploadGuestLists := UploadGuestListsDTO{
		CreateGuests: validatedCreateGuests,
		UpdateGuests: validatedUpdateGuests,
	}
	return &UploadValidationDTO{UploadIssues: uploadIssues, UploadGuestLists: uploadGuestLists}, nil
}

func validateGuest(guest GuestDTO, existingGuestNames []string, isUpdate bool) string {
	if guest.Name == "" {
		return NameMissing
	}

	if guest.GroupNumber == -1 {
		return PartyNumberMissing
	}

	if !isUpdate && common.ArrayContains(existingGuestNames, strings.ToLower(guest.Name)) {
		return DuplicateGuest
	}

	if len(guest.Name) > 100 {
		return NameTooLong
	}

	if guest.Email != "" && len(guest.Email) > 50 {
		return EmailTooLong
	}

	if guest.Phone != "" && len(guest.Phone) > 50 {
		return PhoneTooLong
	}

	return ""
}

func (s *guestService) GenerateCSVContent(params map[string]string) (string, *common.CustomError) {
	guestDtos, err := s.GetAll(params)
	if err != nil {
		return "", err
	}
	var csvBuilder strings.Builder
	writer := csv.NewWriter(&csvBuilder)

	headers := CsvHeaders()
	if err := writer.Write(headers); err != nil {
		return "", common.NewCustomError(500, "Error writing headers to csv file", err)
	}

	for _, guestDto := range guestDtos {
		row := []string{
			guestDto.Id,
			guestDto.EventId,
			guestDto.Name,
			fmt.Sprint(guestDto.GroupNumber),
			guestDto.SerialNumber,
		}
		//SubEvents
		var subEventNames []string
		for _, subevent := range guestDto.SubEvents {
			subEventNames = append(subEventNames, subevent.Name)
		}
		row = append(row, strings.Join(subEventNames, ", "))
		//Email
		row = append(row, guestDto.Email)
		//Accommodation
		row = append(row, guestDto.Accommodation.Name)
		row = append(row, guestDto.Accommodation.RoomNumber)
		//Dietary Restrictions
		row = append(row, guestDto.DietaryRestrictions)
		//Drinks
		row = append(row, fmt.Sprint(guestDto.Drinks.WillDrinkAlcohol))
		row = append(row, strings.Join(guestDto.Drinks.Preferences, ", "))
		//Arrival
		row = append(row, guestDto.Arrival.Type)
		arrivalDateTime := getDateTime(guestDto.Arrival.Time)
		row = append(row, arrivalDateTime[0])
		row = append(row, arrivalDateTime[1])
		row = append(row, guestDto.Arrival.Number)
		//Departure
		row = append(row, guestDto.Departure.Type)
		departureDateTime := getDateTime(guestDto.Departure.Time)
		row = append(row, departureDateTime[0])
		row = append(row, departureDateTime[1])
		row = append(row, guestDto.Departure.Number)
		if err := writer.Write(row); err != nil {
			return "", common.NewCustomError(500, "Error writing rows to csv file", err)
		}
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		return "", common.NewCustomError(500, "Error flushing csv file", err)
	}
	return csvBuilder.String(), nil
}

func getDateTime(dateTime string) []string {
	if dateTime == "" {
		return []string{"", ""}
	}
	var dateTimeSplit = strings.Split(dateTime, "T")

	dateLayout := "2006-01-02"

	parsedTime, err := time.Parse(dateLayout, dateTimeSplit[0])
	if err != nil {
		dateTimeSplit[0] = "ERROR"
		return dateTimeSplit
	}

	dateTimeSplit[0] = parsedTime.Format("01/02/2006")

	return dateTimeSplit

}

func (s *guestService) UploadGuests(params map[string]string, file multipart.File, ext string) (*UploadGuestListsDTO, *common.CustomError) {
	eventId, ok := params["eventId"]
	if !ok || eventId == "" {
		return nil, common.NewCustomError(400, "Event ID not provided", nil)
	}

	subevents, err := s.SubEventService.GetAll(params)
	if err != nil {
		return nil, err
	}
	accommodations, err := s.AccommodationService.GetAll(params)
	if err != nil {
		return nil, err
	}

	var uploadGuestLists *UploadGuestListsDTO
	switch ext {
	case ".csv":
		uploadGuestLists, err = parseCSV(file, eventId, subevents, accommodations)
		if err != nil {
			return nil, err
		}
	case ".xlsx":
		uploadGuestLists, err = parseXLSX(file, eventId, subevents, accommodations)
		if err != nil {
			return nil, err
		}
	default:
		return nil, common.NewCustomError(400, "Unsupported file type", nil)
	}

	return uploadGuestLists, nil
}

func parseCSV(r io.Reader, eventId string, subEvents []*subevents.SubEventDTO, accommodations []*accommodation.AccommodationDTO) (*UploadGuestListsDTO, *common.CustomError) {
	reader := csv.NewReader(r)
	headers, err := reader.Read()
	if err != nil {
		return nil, common.NewCustomError(500, "Failed to read csv header", nil)
	}
	headerMap := make(map[string]int)
	for i, h := range headers {
		headerMap[h] = i
	}

	var createGuests, updateGuests []GuestDTO

	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, common.NewCustomError(500, "Failed to read csv record", nil)
		}

		rowData := make(map[string]string)
		for key, idx := range headerMap {
			if idx < len(record) {
				rowData[key] = record[idx]
			}
		}

		if id, ok := rowData["Id"]; ok && id != "" {
			guest, err := updateGuestFromData(rowData, eventId, subEvents, accommodations)
			if err != nil {
				return nil, err
			}
			updateGuests = append(updateGuests, guest)
		} else {
			guest, err := createGuestFromData(rowData, eventId, subEvents, accommodations)
			if err != nil {
				return nil, err
			}
			createGuests = append(createGuests, guest)
		}
	}

	return &UploadGuestListsDTO{
		CreateGuests: createGuests,
		UpdateGuests: updateGuests,
	}, nil
}

func parseXLSX(r io.Reader, eventId string, subEvents []*subevents.SubEventDTO, accommodations []*accommodation.AccommodationDTO) (*UploadGuestListsDTO, *common.CustomError) {
	dataBytes, err := io.ReadAll(r)
	if err != nil {
		return nil, common.NewCustomError(500, "Failed to read XLSX file", err)
	}

	f, err := excelize.OpenReader(strings.NewReader(string(dataBytes)))
	if err != nil {
		return nil, common.NewCustomError(500, "Failed to open XLSX file", err)
	}
	defer f.Close()

	sheetName := f.GetSheetName(0)
	if sheetName == "" {
		return nil, common.NewCustomError(500, "No sheet found in the XLSX file", nil)
	}

	rows, err := f.GetRows(sheetName)
	if err != nil {
		return nil, common.NewCustomError(500, "Failed to read rows from XLSX", err)
	}

	if len(rows) < 2 {
		return nil, common.NewCustomError(400, "XLSX file does not contain enough rows", nil)
	}

	headerMap := make(map[string]int)
	headers := rows[0]
	for i, h := range headers {
		headerMap[h] = i
	}

	var createGuests, updateGuests []GuestDTO

	for _, record := range rows[1:] {
		rowData := make(map[string]string)

		for key, idx := range headerMap {
			if idx < len(record) {
				rowData[key] = record[idx]
			}
		}

		if id, ok := rowData["Id"]; ok && id != "" {
			guest, err := updateGuestFromData(rowData, eventId, subEvents, accommodations)
			if err != nil {
				return nil, err
			}
			updateGuests = append(updateGuests, guest)
		} else {
			guest, err := createGuestFromData(rowData, eventId, subEvents, accommodations)
			if err != nil {
				return nil, err
			}
			createGuests = append(createGuests, guest)
		}
	}

	return &UploadGuestListsDTO{
		CreateGuests: createGuests,
		UpdateGuests: updateGuests,
	}, nil
}

func createGuestFromData(data map[string]string, eventId string, allSubEvents []*subevents.SubEventDTO, allAccommodations []*accommodation.AccommodationDTO) (GuestDTO, *common.CustomError) {
	createdGuest := GuestDTO{
		EventId:             eventId,
		Name:                parseStringField(data["Name"]),
		Email:               parseStringField(data["Email"]),
		Phone:               cleanPhoneField(data["Phone"]),
		SerialNumber:        parseStringField(data["Serial Number"]),
		DietaryRestrictions: parseStringField(data["Dietary Restrictions"]),
	}

	setGuestFields(data, &createdGuest, allSubEvents, allAccommodations)

	return createdGuest, nil
}

func updateGuestFromData(data map[string]string, eventId string, allSubEvents []*subevents.SubEventDTO, allAccommodations []*accommodation.AccommodationDTO) (GuestDTO, *common.CustomError) {
	updatedGuest := GuestDTO{
		Id:                  parseStringField(data["Id"]),
		EventId:             eventId,
		Name:                parseStringField(data["Name"]),
		Email:               parseStringField(data["Email"]),
		Phone:               cleanPhoneField(data["Phone"]),
		SerialNumber:        parseStringField(data["Serial Number"]),
		DietaryRestrictions: parseStringField(data["Dietary Restrictions"]),
	}

	setGuestFields(data, &updatedGuest, allSubEvents, allAccommodations)

	return updatedGuest, nil
}

func setGuestFields(data map[string]string, guest *GuestDTO, allSubEvents []*subevents.SubEventDTO, allAccommodations []*accommodation.AccommodationDTO) *common.CustomError {
	// Group Number
	groupNumber, err := strconv.Atoi(parseStringField(data["Party Number"]))
	if err != nil {
		logger.Global.Error("Error converting Group Number to integer: %w", err)
	}
	guest.GroupNumber = groupNumber

	// Sub Events
	guest.SubEvents = parseSubEvents(data, allSubEvents)

	// Arrival Transportation
	guest.Arrival = parseTransportation(data, true)

	// Departure Transportation
	guest.Departure = parseTransportation(data, false)

	// Accommodation
	guest.Accommodation = parseAccommodation(data, allAccommodations)

	// Drinks
	guest.Drinks = parseDrinks(data)

	return nil
}

func parseStringField(input string) string {
	if input == "" {
		return ""
	}
	return strings.TrimSpace(input)
}

func cleanPhoneField(phone string) string {
	if phone == "" {
		return ""
	}
	trimmed := strings.TrimSpace(phone)
	re := regexp.MustCompile(`[^\d+]`)

	cleaned := re.ReplaceAllString(trimmed, "")

	if strings.Index(cleaned, "+") > 0 {
		cleaned = strings.ReplaceAll(cleaned, "+", "")
	}
	return cleaned
}

func parseSubEvents(data map[string]string, allSubEvents []*subevents.SubEventDTO) []subevents.SubEventDTO {
	var guestSubEvents []subevents.SubEventDTO

	for _, subEvent := range allSubEvents {
		invitedStatus := strings.ToLower(parseStringField(data[subEvent.Name]))
		if invitedStatus != "" && common.ArrayContains([]string{"yes", "done", "y", "true"}, invitedStatus) {
			guestSubEvents = append(guestSubEvents, *subEvent)
		}
	}
	return guestSubEvents

}

func parseTransportation(data map[string]string, isArrival bool) transportation.TransportationDTO {

	var transportation transportation.TransportationDTO

	var prefix string
	if isArrival {
		transportation.IsArrival = true
		prefix = "Arr"
		transportation.Type = parseStringField(data["Arrival Type"])
	} else {
		transportation.IsArrival = false
		prefix = "Dep"
		transportation.Type = parseStringField(data["Departure Type"])
	}

	transportation.Number = parseStringField(data[fmt.Sprintf("%s.Num", prefix)])
	date := parseStringField(data[fmt.Sprintf("%s.Date", prefix)])
	time := parseStringField(data[fmt.Sprintf("%s.Time", prefix)])
	dateTime, err := combineDateTime(date, time)
	if err != nil {
		transportation.Time = ""
	} else {
		transportation.Time = dateTime
	}

	return transportation
}

func combineDateTime(dateStr, timeStr string) (string, *common.CustomError) {
	layout := "01/02/2006 15:04"
	combinedStr := fmt.Sprintf("%s %s", dateStr, timeStr)

	parsedTime, err := time.Parse(layout, combinedStr)
	if err != nil {
		return "", common.NewCustomError(500, "error parsing date and time", err)
	}

	return parsedTime.Format("2006-01-02T15:04:05"), nil
}

func parseAccommodation(data map[string]string, allAccommodations []*accommodation.AccommodationDTO) accommodation.AccommodationDTO {
	accommodationName := parseStringField(data["Accommodation"])
	roomNumber := parseStringField(data["Room Number"])

	var assignedAccommodation accommodation.AccommodationDTO
	for _, accommodation := range allAccommodations {
		if accommodation.Name == accommodationName {
			assignedAccommodation = *accommodation
		}
	}
	assignedAccommodation.RoomNumber = roomNumber
	return assignedAccommodation
}

func parseDrinks(data map[string]string) drinks.DrinksDTO {
	var drinks drinks.DrinksDTO
	willDrink := strings.ToLower(parseStringField(data["Drinks?"]))
	if willDrink != "" && common.ArrayContains([]string{"yes", "done", "y", "true"}, willDrink) {
		drinks.WillDrinkAlcohol = true
	} else {
		drinks.WillDrinkAlcohol = false
	}
	var drinkPreferences []string
	drinkPreferenceString := parseStringField(data["Type of Drink"])
	drinkPreferences = strings.Split(drinkPreferenceString, ";")
	drinks.Preferences = drinkPreferences
	return drinks
}
